apps/withdraw-balancer-v2/src/utils/readUserBalances.ts

```diff
@@ -0,0 +1,32 @@
+import { BigNumber } from "ethers";
+import { type Address, type PublicClient, erc20Abi } from "viem";
+
+export async function readUserBalances(
+  userAddress: string,
+  contracts: string[],
+  client: PublicClient,
+): Promise<BigNumber[]> {
+  try {
+    const calls = contracts.map((contractAddress) => ({
+      address: contractAddress as Address,
+      abi: erc20Abi,
+      functionName: "balanceOf",
+      args: [userAddress as Address],
+    }));
+
+    const balances = await client.multicall({
+      contracts: calls,
+      allowFailure: true,
+    });
+
+    // Filter out any failed calls and map to return bigint array
+    return balances.map((result) =>
+      result.status === "success"
+        ? BigNumber.from(result.result)
+        : BigNumber.from("0"),
+    );
+  } catch (error) {
+    console.error("Error reading balances:", error);
+    return new Array(contracts.length).fill(BigInt(0));
+  }
+}
```

Comment:
I think that you can use same function that we used to read the other token information. 

Also, this function doesn't takes the balancesDiif (due previous hooks) into account.

Reply:
With the last update, I refactored it to read multiple info from many contracts, so It got a little more complex.

About the balanceDiff, I have my doubts if would be worth to consider it... I mean, it seems the chances are pretty small of someone's LP tokens change before a pre-hook, isn't it?